### PR TITLE
fix: add `null` check in `deepFindPathToProperty` function

### LINKED ISSUE #58: [BUG]: deepFindPathToProperty() throws TypeError if graphql response field is null
### What happened?

Some times a graphql response fields have value null. This causes currentValue.hasOwnProperty(searchProp) to throw `TypeError: Cannot read properties of null (reading 'hasOwnProperty')` error. 

<details>
<summary>Minimal Reproducible Example</summary>

```javascript
import nock from 'nock';
import { Octokit } from "octokit";
import { paginateGraphql } from "@octokit/plugin-paginate-graphql";

const MyOctokit = Octokit.plugin(paginateGraphql);
const octokit = new MyOctokit({ auth: "secret123" });


const myQuery = `{ 
    query {
      user(login: "stergion") {
        name
        login
        contributionsCollection {
          earliestRestrictedContributionDate
          latestRestrictedContributionDate
        }
      }
    }
  }`;

nock.disableNetConnect();
nock('https://api.github.com')
  .post('/graphql', '{"query":"{ \\n    query {\\n      user(login: \\"stergion\\") {\\n        name\\n        login\\n        contributionsCollection {\\n          earliestRestrictedContributionDate\\n          latestRestrictedContributionDate\\n        }\\n      }\\n    }\\n  }"}')
  .reply(200, {
    "data": {
      "user": {
        "name": null,
        "login": "stergion",
        "contributionsCollection": {
          "earliestRestrictedContributionDate": "2022-02-13",
          "latestRestrictedContributionDate": "2023-02-14"
        }
      }
    }
  });

// const response = await octokit.graphql.paginate(myQuery);

try {
  const response = await octokit.graphql.paginate(myQuery);
  console.log(response);
} catch (error) {
  console.log(`${error.name}: ${error.status}`);
  console.log(error);
}
```

</details>

### Versions

@octokit/plugin-paginate-graphql@2.0.1
octokit@2.0.14
Node v16.17.0

### Relevant log output

```
TypeError: undefined
TypeError: Cannot read properties of null (reading 'hasOwnProperty')
    at deepFindPathToProperty (/home/mre/node_modules/@octokit/plugin-paginate-graphql/dist-node/index.js:52:22)
    at deepFindPathToProperty (/home/mre/node_modules/@octokit/plugin-paginate-graphql/dist-node/index.js:57:22)
    at findPaginatedResourcePath (/home/mre/node_modules/@octokit/plugin-paginate-graphql/dist-node/index.js:38:33)
    at extractPageInfos (/home/mre/node_modules/@octokit/plugin-paginate-graphql/dist-node/index.js:94:24)
    at Object.next (/home/mre/node_modules/@octokit/plugin-paginate-graphql/dist-node/index.js:122:35)
    at async Function.paginate (/home/mre/node_modules/@octokit/plugin-paginate-graphql/dist-node/index.js:178:22)
    at async file:///home/rme/index.mjs:41:20
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by igwejk**
Resolves #58 

----

### Before the change?
The `deepFindPathToProperty` function would occasionally fail with `TypeError: Cannot read properties of null (reading 'hasOwnProperty')`.

### After the change?

The error has been fixed:
* <a href="diffhunk://#diff-be5e0092082abfc624a87b5ab130c1f22c2c68205637088e88268bbcc776e96dL7-R104">`src/object-helpers.ts`</a>: The `deepFindPathToProperty` function was refactored and its return type updated from `string[]` to `string[] | null`. The refactoring includes the introduction of a new type `TreeNode` and several new helper functions like `getDirectPropertyPath`, `makeTreeNodeChildrenFromData`, and `findPathToObjectContainingProperty`. The `findPaginatedResourcePath` function was also updated to accommodate these changes.
* <a href="diffhunk://#diff-f0e1dffba6fc83b2e2bfe81aeab3e340f2b50b5cbf6aac225bfc5dc2536ea4eaL40-R40">`test/typescript-validate.ts`</a>: The `pageInfoBackwaredExported` function was renamed to `pageInfoBackwardExported` for correct spelling.
* <a href="diffhunk://#diff-85d9d85b851fb0a476af9b2bc59226f08f174e1d5d8a98f1e7c855fee91db97fR1-R73">`test/object-helpers.test.ts`</a>: New tests were added for the `findPaginatedResourcePath` function to validate its behavior in different scenarios. 

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [x] No

----

### CODE REVIEW

**Comment by davelosert** [comment]
Hey @igwejk, thank you very much for the PR and for trying to fix #58!

Could you elaborate a bit more as to why you chose to entirely reimplement the `deepFindPathToProperty`-Function with a lot more code than we had previously, given that you were only trying to fix errors with possible null-values of certain fields?

It seems like you are covering other use-cases with this implementation as well and I don't quite understand 'which' and 'why' yet. 🙂

**Comment by davelosert** [comment]
Hey @igwejk,

as already discussed internally, we've chosen to not support multiple (or nested) pagination with this library on purpose, as this can lead to all sort of problems (you can find more info and a link to a more in depth-explanation in the [README Section: Unsupported Nested Pagination](https://github.com/octokit/plugin-paginate-graphql.js?tab=readme-ov-file#unsupported-nested-pagination).

As I also don't see a GraphQL Response that would be so nested that the current, recursive implementation would cause a overflowing call stack, I'd rather keep the current, more lightweight implementation and just use a null-check to avoid the error reported in #58.

Still, thank you again for the contribution and your work here! 👍🏻

**Comment by karpikpl** [comment]
> Hey @igwejk,
> 
> as already discussed internally, we've chosen to not support multiple (or nested) pagination with this library on purpose, as this can lead to all sort of problems (you can find more info and a link to a more in depth-explanation in the [README Section: Unsupported Nested Pagination](https://github.com/octokit/plugin-paginate-graphql.js?tab=readme-ov-file#unsupported-nested-pagination).
> 
> As I also don't see a GraphQL Response that would be so nested that the current, recursive implementation would cause a overflowing call stack, I'd rather keep the current, more lightweight implementation and just use a null-check to avoid the error reported in #58.
> 
> Still, thank you again for the contribution and your work here! 👍🏻

Was that null check ever implemented?

**Comment by igwejk** [comment]
@karpikpl It was fixed with with an addition of support for deep nested `pageInfo` which was not required. I've now updated the PR to keep only the null check.

**Review by wolfy1339 — APPROVED** [review]

File: src/object-helpers.ts [hunk 1]
```diff
@@ -4,11 +4,11 @@ const isObject = (value: any) =>
   Object.prototype.toString.call(value) === "[object Object]";
 
 function findPaginatedResourcePath(responseData: any): string[] {
-  const paginatedResourcePath = deepFindPathToProperty(
+  const paginatedResourcePath: string[] | null = deepFindPathToProperty(
     responseData,
     "pageInfo",
   );
-  if (paginatedResourcePath.length === 0) {
+  if (paginatedResourcePath === null || paginatedResourcePath.length === 0) {
```
Comment (kachick) [inline src/object-helpers.ts:11]: Did this line fix #58? The stacktrace and the description looks like it happens in `deepFindPathToProperty` implementation, not for the returned value.

>This causes currentValue.hasOwnProperty(searchProp) to throw TypeError: Cannot read properties of null (reading 'hasOwnProperty') error.

https://github.com/octokit/plugin-paginate-graphql.js/blob/17e58b6c3ec476e3877a9994b3b9f68a9ea6d8bf/src/object-helpers.ts#L26-L28

Comment (kachick) [reply src/object-helpers.ts:11]: https://github.com/octokit/plugin-paginate-graphql.js/pull/59/files#diff-be5e0092082abfc624a87b5ab130c1f22c2c68205637088e88268bbcc776e96dR26 looks focusing to there

**Pull request merged**